Example VRF contracts

alexroan opened the pull request:
Changes:
- New `examples/` folder
- Move VRFD20 to examples
- Modify VRFD20
- Testing

se3000 commented on evm-contracts/src/v0.6/examples/VRFBasic.sol:
```diff
@@ -0,0 +1,71 @@
+pragma solidity 0.6.6;
```
I vote to cut this contract. It doesn't really add much clarity, the D20 is already relatively simple. This just means we have twice as much example code to maintain moving forward. What do you think?

alexroan replied:
As a learning tool, I think it's great to have. It essentially mimics our [VRF docs example](https://docs.chain.link/docs/get-a-random-number). 

If our main source of documentation was _this repo_, then I would strongly vote in favour of keeping this contract.

However, we have external docs for that, so I'm happy to remove this.

se3000 commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,112 @@
+contract VRFD20 is VRFConsumerBase {
+    using SafeMathChainlink for uint;
+
+    struct Roll {
```
There is no reason to keep the request ID around, that is really just internal plumbing details. I think we should not expose it in the example, as it may give the impression to other people that they need to expose it with the randomness as well.

se3000 commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,112 @@
+     *
+     * @return requestId bytes32
+     */
+    function currentRollRequest() public view returns (bytes32 requestId) {
```
Similarly, I don't see much value in exposing the request ID. I guess in this case, you can see if a new roll is happening. Still, there is a bit of a disconnect, you get a request ID here, but then you look up results based on the Roll's index. I recommend just exposing index to the end user.

se3000 commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,138 @@
+ * of a 20 sided die
+ */
+contract VRFD20 is VRFConsumerBase, Owned {
+    using SafeMathChainlink for uint;
```
```suggestion
```
I believe the SafeMath behavior is already inherited from VRFConsumerBase. I believe we can remove this with zero consequences.

se3000 replied:
Side note: I think we should always prefer the explicit `uint256` over the implied `uint`. This has bitten me before when calculating function sigantures.

se3000 commented on evm-contracts/test/v0.6/VRFD20.test.ts:
```diff
@@ -0,0 +1,251 @@
+  describe('#latestResult', () => {
+    it('reverts when there are no results', async () => {
+      await matchers.evmRevert(async () => {
+        await vrfD20.latestResult()
+      })
```
```suggestion
        await vrfD20.latestResult(),
        'Invalid result number',
      })
```
Definitely worth making sure reverts are happening for the reason you suggest.

se3000 commented on evm-contracts/test/v0.6/VRFD20.test.ts:
```diff
@@ -0,0 +1,251 @@
+  describe('#getResult', () => {
+    it('reverts when a number too high is used', async () => {
+      await matchers.evmRevert(async () => {
+        await vrfD20.getResult(99)
+      })
```
```suggestion
        await vrfD20.getResult(99),
        'Invalid result number',
      })
```

se3000 reviewed (changes requested):
Couple specific little changes. Larger changes:
- It seems like it is expected that LINK is deposited directly into this contract. For the sake of recovering it, I think it would be worth adding a `withdrawLINK(uint256,address)` method that only the owner can call.
- I'm not as certain about this one, but I think we should remove the `rollInProgress` check. I don't think there is any reason you can't actually have multiple rolls running in parallel, but this seems to imply that you should not.

alexroan commented:
> Couple specific little changes. Larger changes:
> 
> * It seems like it is expected that LINK is deposited directly into this contract. For the sake of recovering it, I think it would be worth adding a `withdrawLINK(uint256,address)` method that only the owner can call.
> * I'm not as certain about this one, but I think we should remove the `rollInProgress` check. I don't think there is any reason you can't actually have multiple rolls running in parallel, but this seems to imply that you should not.

* Added `withdrawLINK` function
* Removed all references to `rollInProgress`
* Added specific messages to all revert tests

alexroan commented:
Change of scope.

- Remove the roll history
- 1:1 roller:roll
- Each number has an associated character type (d&d style)

alexroan commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,118 @@
+
+    mapping(address => bytes32) private s_rollers;
+    mapping(bytes32 => uint256) private s_results;
+
```
@se3000 how does this look?

The requestId needs to be a key for the fulfil function, and the address of the roller need to be a key so that they can query their result/d&d character.

se3000 replied:
I think that on the lookup it would be more efficient to read the address directly instead of pulling the request ID and then using that to look up the address. So it would be:
```
    mapping(address => bytes32) private s_rollers;
    mapping(address => uint256) private s_results;
```
Then when you want to roll, you just make sure that the there is no roll or result present for the address.

se3000 replied:
You could possibly consolidate the above two mappings into a single struct, but I could see the argument that this is simpler too.

alexroan replied:
The problem with your example is the fulfillment is difficult. The address is not involved in the fulfillment process, so setting `s_results` when the dice has landed will involve searching through all address keys in `s_rollers` to find the correct `requestId` value, then update `s_results`.

alexroan replied:
... unless we only allow 1 roll to be in progress at any given time and lock the contract up during it.

alexroan replied:
I've kept this the same as fulfilment will be inefficient otherwise.

se3000 replied:
Good point about retrieving the address on fulfillment, I missed that. Still, the final read, `s_houses[s_results[s_rollers[player]]]`, requires 3 reads from storage and the s_houses read is extra large because it's a string; that's a lot of gas for a read.

How about:
```
mapping(bytes32 => address) private s_requests;
mapping(address => uint256) private s_results;
```
When the roll happens, mark the address's `s_results` with some special number like 42, which is outside the range of dice rolls and just indicates a roll is in progress. Then, create a helper function which just has an array of strings in memory and looks up the string based on index.

se3000 commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,149 @@
+     * @param requestId bytes32
+     * @param randomness The random result returned by the oracle
+     */
+    function fulfillRandomness(bytes32 requestId, uint256 randomness) internal override {
```
I know we normally push internal methods below public, but this is kind of where the meat of the contract lies. I'd put the constructor and `rollDice` before it, and everything else after. That is also kind of inline with the actual usage work-flow.

se3000 commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+     * @return house name string
+     */
+    function getHouseName(uint256 id) private view returns (string memory) {
+        string[20] memory houseNames = [
```
Is it possible to extract these into a constant? This would be a much smaller gas optimization, but I think we'd still save.

alexroan replied:
Nope, due to this from solc: `TypeError: Constants of non-value type not yet implemented.`

coventry commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+
+    /**
+     * @notice Requests randomness from a user-provided seed
+     * @dev This is only an example implementation and not necessarily suitable for mainnet.
```
I would move this to the header comment for the entire contract. Otherwise, people might construe it to only apply to this method.

coventry commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+     * @notice Callback function used by VRF Coordinator to return the random number
+     * to this contract.
+     * @dev This is where you do something with randomness!
+     * @dev The VRF Coordinator will only send this function verified responses.
```
I would add something like "..., and the parent `VRFConsumerBase` contract ensures that this method only receives randomness from the designated `VRFCoordinator`."

coventry commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+     * @param randomness The random result returned by the oracle
+     */
+    function fulfillRandomness(bytes32 requestId, uint256 randomness) internal override {
+        uint256 result = randomness.mod(20).add(1);
```
I would call this `d20Value` or something similar.

coventry commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+     *
+     * @param requestId bytes32
+     * @param randomness The random result returned by the oracle
+     */
```
We should add an `@dev` note here explaining that some action on the contract state should be taken here, like storing the result, and a warning to  take care to avoid having multiple VRF requests in flight if their order of arrival would result in contract states with different outcomes. Otherwise miners or the VRF operator would could take advantage by controlling the order.

coventry commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+     * @notice Requests randomness from a user-provided seed
+     * @dev This is only an example implementation and not necessarily suitable for mainnet.
+     * @dev You must review your implementation details with extreme care.
+     *
```
I would add a note warning that if the VRF response is delayed, they should avoid calling `requestRandomness` repeatedly, as that would give miners/VRF operators latitude about which VRF response arrives first.

coventry commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * @notice Withdraw LINK from this contract.
+     * @param to the address to withdraw LINK to
+     * @param value the amount of LINK to withdraw
```
I would add a note here that this is an example only, and in a real contract withdrawals should happen according to the [established withdrawal pattern](https://docs.soliditylang.org/en/v0.4.24/common-patterns.html#withdrawal-from-contracts), to protect users from burning their funds by fat-fingering the `to` address.

coventry commented on evm-contracts/src/v0.6/examples/VRFD20.sol:
```diff
@@ -0,0 +1,164 @@
+     * @return house name string
+     */
+    function getHouseName(uint256 id) private view returns (string memory) {
+        string[20] memory houseNames = [
```
This disregards the Free Folk, who have no house. We're going to need you to apologize on #gameofthrones Twitter.😊

alexroan replied:
The savages have no place in Westeros

coventry commented on evm-contracts/src/v0.6/tests/VRFTestnetD20.sol:
```diff
@@ -1,53 +0,0 @@
-pragma solidity 0.6.6;
```
It's fine to get rid of this, but I'm surprised that the golang [TestD20ExampleContract](https://github.com/smartcontractkit/chainlink/blob/feature%2F%23175865202-d20-example/core/services/vrf/vrf_testnet_d20_test.go) is still passing.

coventry commented on evm-contracts/src/v0.6/tests/VRFCoordinatorMock.sol:
```diff
@@ -0,0 +1,40 @@
+pragma solidity 0.6.6;
```
This looks sane, but we probably want an integration test at some point (not in this PR) like [TestD20ExampleContract](https://github.com/smartcontractkit/chainlink/blob/feature%2F%23175865202-d20-example/core/services/vrf/vrf_testnet_d20_test.go).

coventry commented on evm-contracts/test/v0.6/VRFD20.test.ts:
```diff
@@ -0,0 +1,290 @@
+
+  it('has a limited public interface', () => {
+    matchers.publicAbi(vrfD20Factory, [
+      // Owned
```
Why are the commented-out attributes missing?

se3000 replied:
That's a pattern in the tests, it's just to differentiate where in the inherritance tree these public methods come from.

alexroan replied:
Yeah, they aren't attributes, @coventry. They're comments describing which contracts the subsequent attributes are inherited (and exposed) from.

coventry commented on evm-contracts/test/v0.6/VRFD20.test.ts:
```diff
@@ -0,0 +1,290 @@
+    })
+  })
+
+  describe('#hosue', () => {
```
typo

coventry reviewed (changes requested):
The overall contract looks good, but I've made some suggestions for more documentation.

I think at some point, we should probably hook this into an integration test to ensure its correct operation in the full system.
